reference/test/preconditioner/isai_kernels.cpp

```diff
@@ -0,0 +1,626 @@
+    using Composition = typename TestFixture::Composition;
+    using value_type = typename TestFixture::value_type;
+    const auto comp = gko::share(Composition::create(
+        gko::share(this->l_sparse), gko::share(this->u_sparse)));
+    const auto isai = this->isai_factory->generate(comp);
+
+    auto l_inv = isai->get_approx_inverse_l();
+    auto u_inv = isai->get_approx_inverse_u();
+
+    GKO_ASSERT_MTX_NEAR(l_inv, this->l_sparse_inv, r<value_type>::value);
+    GKO_ASSERT_MTX_NEAR(u_inv, this->u_sparse_inv, r<value_type>::value);
+}
+
+
+TYPED_TEST(Isai, GeneratesWithUnsortedCsr)
+{
+    using Csr = typename TestFixture::Csr;
+    using Composition = typename TestFixture::Composition;
+    using value_type = typename TestFixture::value_type;
+    const auto comp = gko::share(Composition::create(
+        gko::share(this->l_s_unsorted), gko::share(this->u_s_unsorted)));
+
+    const auto isai = this->isai_factory->generate(comp);
+    auto l_inv = isai->get_approx_inverse_l();
+    auto u_inv = isai->get_approx_inverse_u();
+
+    GKO_ASSERT_MTX_NEAR(l_inv, this->l_sparse_inv, r<value_type>::value);
+    GKO_ASSERT_MTX_NEAR(u_inv, this->u_sparse_inv, r<value_type>::value);
+}
+
+
+TYPED_TEST(Isai, ApplyMultipleRhs)
+{
+    using Dense = typename TestFixture::Dense;
+    using Composition = typename TestFixture::Composition;
+    using T = typename TestFixture::value_type;
+    auto x = Dense::create(this->exec, gko::dim<2>{3, 3});
+    auto result = Dense::create_with_config_of(gko::lend(x));
+    this->l_dense->apply(gko::lend(this->u_dense), gko::lend(x));
+    const auto comp = gko::share(Composition::create(
+        gko::share(this->l_dense), gko::share(this->u_dense)));
+    const auto isai = this->isai_factory->generate(comp);
+
+    isai->apply(gko::lend(x), gko::lend(result));
+
+    // Testing for: U^-1 * L^-1 * L * U == I
+    GKO_ASSERT_MTX_NEAR(
+        result, l({I<T>{1, 0, 0}, I<T>{0, 1, 0}, I<T>{0, 0, 1}}), r<T>::value);
+}
+
+
+TYPED_TEST(Isai, ApplyWithCacheReset)
+{
+    using Dense = typename TestFixture::Dense;
+    using Composition = typename TestFixture::Composition;
+    using T = typename TestFixture::value_type;
+    const auto vec1 = gko::initialize<Dense>({128, -64, 32}, this->exec);
+    const auto vec2 = gko::initialize<Dense>(
+        {I<T>{16, 64}, I<T>{32, 8}, I<T>{-16, 128}}, this->exec);
+    auto res1 = Dense::create(this->exec, vec1->get_size());
+    auto res2 = Dense::create(this->exec, vec2->get_size());
+    const auto comp = gko::share(Composition::create(
+        gko::share(this->l_dense), gko::share(this->u_dense)));
+    const auto isai = this->isai_factory->generate(comp);
+
+    isai->apply(gko::lend(vec1), gko::lend(res1));
+    isai->apply(gko::lend(vec2), gko::lend(res2));
+
+    GKO_ASSERT_MTX_NEAR(res1, l(I<T>{25, -40, -4}), r<T>::value);
+    GKO_ASSERT_MTX_NEAR(res2,
+                        l({I<T>{.625, 14.125}, I<T>{5, -43}, I<T>{-.5, -18.5}}),
```

Comment:
So far this one seems like the only one (with vec2 up here) where the `I<T>` should be needed because there can be list of complex of T vs list of list of T ambiguity.

Reply:
Unfortunately, I also need the `l()` here (which is defined in [`core/test/utils/assertions.hpp`](https://github.com/ginkgo-project/ginkgo/blob/develop/core/test/utils/assertions.hpp#L649) as a function taking and returning an `initializer_list`). I need it so the macro does not read the commas `,` as more arguments.

By the way: I did some testing, and we could use this `l(std::initializer_list<T> )` function instead of the alias `I<T>` (`using I<T> = std::initializer_list<T>`). The draw back, however, would be that we need an additional round brace in the end.
```c++
I<T>{1, 2, 3};
l<T>({1, 2, 3}); // equivalent to statement above
```
Personally, I would keep both variants (the alias is shorter when we need to specify the template argument), but I thought it is good to know what the `l()` function actually does.